File: solver/cookies.js
// Cookie scoping for the solved session.
//
// browser.cookies() takes no arguments: it returns every cookie in the profile,
// for every origin the session touched. The URL that used to be passed to it was
// silently ignored, which broke two things at once.
//
//   - A challenge run navigates cross-origin and back, and the challenge host
//     sets its own cf_clearance on the way through. The clearance check
//     therefore accepted another domain's cookie as proof that the target had
//     been solved.
//   - The `cookies` string handed back is meant to be pasted into a Cookie:
//     header for the target. Built from every origin, it sends foreign cookies
//     to the target, and duplicate names collapse in an order nothing controls.
//
// Both measured against Chromium 141. browser.cookies() and browser.cookies(url)
// return byte-identical lists, and a solve driven through a cross-origin
// redirect chain produced this before the fix:
//
//   cookies:  cf_clearance=THIRD_PARTY; tp_junk=1; cf_clearance=REAL; __cf_bm=bm1
//   reported: cf_clearance=THIRD_PARTY
//
// — two clearance cookies in one header, the wrong one first, and the wrong one
// reported as the solve.
//
// page.cookies(url) applies real domain and path matching in the browser, which
// is exactly the question being asked, so it is the primary path. The matcher
// below is the fallback for puppeteer builds where the page-level method is
// gone, and it deliberately errs toward excluding a cookie: sending one cookie
// too few costs a retry, sending one too many leaks it to the wrong host.

// targetScope reduces a URL to what cookie matching needs. Throws on a URL that
// does not parse, which is a caller bug rather than a cookie mismatch.
export function targetScope(target) {
  const u = new URL(target);
  return { host: u.hostname.toLowerCase(), path: u.pathname || "/" };
}

// cookieInScope implements the domain-match and path-match rules from RFC 6265
// §5.1.3-5.1.4, minus the public-suffix check the browser does for us on the
// primary path.
//
// A leading dot on the domain attribute is legacy syntax for "and subdomains",
// which is also what a bare host-matching cookie means to a subdomain, so both
// are treated the same: exact host, or a dot-boundary suffix of it. Matching on
// a bare suffix would let "evil-example.com" collect "example.com" cookies.
export function cookieInScope(cookie, scope) {
  const domain = String(cookie.domain || "")
    .replace(/^\./, "")
    .toLowerCase();
  if (!domain) return false;
  if (scope.host !== domain && !scope.host.endsWith("." + domain)) return false;

  const cookiePath = cookie.path || "/";
  if (cookiePath === "/") return true;
  if (scope.path === cookiePath) return true;
  // "/admin" covers "/admin/x" but not "/administrator".
  return scope.path.startsWith(
    cookiePath.endsWith("/") ? cookiePath : cookiePath + "/"
  );
}

// cookiesForUrl returns only the cookies that belong on a request to target.
//
// The whole-profile jar is the read, and the filter above is what makes it
// correct. page.cookies(url) does the matching in the browser and was the
// primary path here, which was a mistake: puppeteer-real-browser re-wraps pages
// on targetcreated to drive the challenge widget, so the page handle this code
// holds is not always attached to the target that did the solving. When it is
// not, page.cookies returns an empty list — and an empty list is not an error,
// so it was taken as the answer and a solved session reported no cookies at all.
// Against a live Cloudflare challenge that produced `"cookie_list":[]` for a
// site that had just issued a cf_clearance.
//
// browser.cookies() reads the profile rather than a page, so it does not depend
// on which handle is live. It only became usable in puppeteer 23.7; the page
// read stays as the fallback for older trees, where it is the only option.
export async function cookiesForUrl(browser, page, target) {
  let scope;
  try {
    scope = targetScope(target);
  } catch {
    return [];
  }

  if (browser && typeof browser.cookies === "function") {
    const all = await browser.cookies().catch(() => null);
    if (all) return all.filter((c) => cookieInScope(c, scope));
  }
  if (page && typeof page.cookies === "function") {
    try {
      return await page.cookies(target);
    } catch {}
  }
  return [];
}

File: internal/solver/cookies.go
package solver

import (
	"net/url"
	"strings"

	"github.com/JSInvasor/Gohttp-clientfingerprintemulateandfastest/internal/cdp"
)

// Cookie scoping for the solved session.
//
// The jar read is the whole context's, and the filter here is what makes it
// correct. Two things go wrong without it, and both did:
//
//   - A challenge run navigates cross-origin and back, and the challenge host
//     sets its own cf_clearance on the way through. The clearance check then
//     accepts another domain's cookie as proof that the target was solved.
//   - The header-ready string is meant to be pasted into a Cookie: header for
//     the target. Built from every origin it sends foreign cookies to the
//     target, and duplicate names collapse in an order nothing controls.
//
// Measured against Chromium 141, a solve driven through a cross-origin redirect
// chain produced this before the fix:
//
//	cookies:  cf_clearance=THIRD_PARTY; tp_junk=1; cf_clearance=REAL; __cf_bm=bm1
//	reported: cf_clearance=THIRD_PARTY
//
// — two clearance cookies in one header, the wrong one first, and the wrong one
// reported as the solve.

// Scope is what cookie matching needs from a URL.
type Scope struct {
	Host string
	Path string
}

// TargetScope reduces a URL to a Scope. It reports whether the URL parsed; a URL
// that does not is a caller bug rather than a cookie mismatch.
func TargetScope(target string) (Scope, bool) {
	u, err := url.Parse(target)
	if err != nil || u.Hostname() == "" {
		return Scope{}, false
	}
	path := u.Path
	if path == "" {
		path = "/"
	}
	return Scope{Host: strings.ToLower(u.Hostname()), Path: path}, true
}

// CookieInScope implements the domain-match and path-match rules from RFC 6265
// §5.1.3-5.1.4, minus the public-suffix check.
//
// A leading dot on the domain attribute is legacy syntax for "and subdomains",
// which is also what a bare host-matching cookie means to a subdomain, so both
// are treated the same: exact host, or a dot-boundary suffix of it. Matching on
// a bare suffix would let "evil-example.com" collect "example.com" cookies.
//
// It deliberately errs toward excluding a cookie: sending one too few costs a
// retry, sending one too many leaks it to the wrong host.
func CookieInScope(c cdp.Cookie, scope Scope) bool {
	domain := strings.ToLower(strings.TrimPrefix(c.Domain, "."))
	if domain == "" {
		return false
	}
	if scope.Host != domain && !strings.HasSuffix(scope.Host, "."+domain) {
		return false
	}

	path := c.Path
	if path == "" || path == "/" {
		return true
	}
	if scope.Path == path {
		return true
	}
	// "/admin" covers "/admin/x" but not "/administrator".
	prefix := path
	if !strings.HasSuffix(prefix, "/") {
		prefix += "/"
	}
	return strings.HasPrefix(scope.Path, prefix)
}

// CookiesForURL returns only the cookies that belong on a request to target.
func CookiesForURL(all []cdp.Cookie, target string) []cdp.Cookie {
	scope, ok := TargetScope(target)
	if !ok {
		return nil
	}
	out := make([]cdp.Cookie, 0, len(all))
	for _, c := range all {
		if CookieInScope(c, scope) {
			out = append(out, c)
		}
	}
	return out
}

// CookieHeader renders cookies the way a Cookie: header carries them.
func CookieHeader(cookies []cdp.Cookie) string {
	parts := make([]string, 0, len(cookies))
	for _, c := range cookies {
		parts = append(parts, c.Name+"="+c.Value)
	}
	return strings.Join(parts, "; ")
}

// HasClearance reports whether this list carries the cookie that decides a
// solve. Everything else in the jar is context.
func HasClearance(cookies []cdp.Cookie) bool {
	for _, c := range cookies {
		if c.Name == "cf_clearance" {
			return true
		}
	}
	return false
}
